### PR TITLE
Mha decomposition update

### PR DESCRIPTION
**Pull request by elad-c**
MHA node decomposition update: implement MHA iterations with layers reuse instead of folding the iterations to batch

### CODE REVIEW

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 1]
```diff
@@ -143,15 +142,79 @@ def _standarize_input_shapes(graph, name, num_input_edges, params):
...
+    @staticmethod
+    def gen_reuse_params(reuse_index, name, reuse_params):
+        """
+        Generate reuse parameters kwrags: REUSE & REUSE_GROUP according to reuse_index.
+
+        :param reuse_index: index representing how many times the node is reused. None for not reused
+        :param name: original node name to calculate the REUSE_GROUP that's renamed each call
+        :param reuse_params: MHA node default params. Defaults to no reuse
+        :return: dictionaty of reuse params
+        """
```
Comment (reuvenperetz) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:154]: Please use Numpy documentation convention

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 2]
```diff
@@ -143,15 +142,79 @@ def _standarize_input_shapes(graph, name, num_input_edges, params):
...
+
+        if reuse_index is None:
+            return reuse_params
+        else:
+            return {REUSE: reuse_index != 0,
+                    REUSE_GROUP: name if reuse_index == 0 else f'{name}_reused_{reuse_index}'
+                    }
```
Comment (reuvenperetz) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:161]: Please use REUSED_IDENTIFIER from node_builder

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 3]
```diff
@@ -143,15 +142,79 @@ def _standarize_input_shapes(graph, name, num_input_edges, params):
...
+
+    @staticmethod
+    def get_reuse_suffix(reuse_index):
+        """
+        Generate node name suffix according to reuse naming
+        :param reuse_index: index representing how many times the node is reused. None for not reused
+        :return: string suffix to be concatenated to resued node name
+        """
+        return '' if reuse_index == 0 or reuse_index is None else f'_reused_{reuse_index}'
+
```
Comment (reuvenperetz) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:171]: Same (REUSED_IDENTIFIER...)

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 4]
```diff
@@ -143,15 +142,79 @@ def _standarize_input_shapes(graph, name, num_input_edges, params):
...
+    def _calc_attention_head(self, graph, q_in_node, k_in_node, v_in_node, mha_node,
+                             iter_index, head_index, params):
         """
         Generate the attention head subgraph: Dot(softmax(Dot(proj(Q), proj(K))), proj(V))
         :param graph: input graph
         :param q_reshape_node:
         :param k_reshape_node:
         :param v_reshape_node:
         :param mha_node: MHA node
+        :param iter_index: iteration index
```
Comment (reuvenperetz) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:217]: Please complete missing arguments

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py (hunk 1 above)
Comment (elad-c) [reply model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:162]: Google convention :)

**Review by reuvenperetz — APPROVED** [review]

**Pull request merged**